### PR TITLE
Client base options on Azure Core

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: #1732 

This PR create ClientOptions on Azure Core and updates KeyVault keys clientOptions to derive from it.
Azure Core exposes a constructor for a new pipeline which takes the clientOptions base class, per service operation and retry policies and construct the pipeline. This constructor ensures that all the expected policies are in the pipeline. 

The PR disables Storage project for now until it gets updated. @Jinming-Hu  please comment if you want to add storage changes here or in a new PR

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:49]: why do we need all these constructors, and member functions? Why not just

```
struct ClientOptions {
    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perOperationPolicies;
    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perRetryPolicies;
    Azure::Core::Http::TransportPolicyOptions Transport;
    ...
};
```

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:54]: So when you make a copy of `ClientOptons`, only policies are preserved, all other fields are just abandoned?

File: CMakeLists.txt
Comment (Jinming-Hu) [inline CMakeLists.txt:96]: if storage sdk doesn't build, you can just comment it out. You don't have to make all the changes to storage sdk

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:152]: So we decided to drop the support for retry to secondary

File: CMakeLists.txt
Comment (vhvb1989) [reply CMakeLists.txt:96]: Yes, I just made a few changes as examples, I expect it to be dismissed and storage team would decide how to fix

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:49]: The per operation and per retry are private, once they are added we don't want them public

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:54]: Yes, to create the pipeline, Core only cares about the client options in the base class.
All other fields are used by each service sdk to create the list of policies that are passed to core to be added to the pipeline

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:152]: No, it just need to be re implemented as a policy after the core retry policy which will check the request the retry number for the that request to decide if switching to secondary

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:49]: I don't quite understand why we need to do that. To me, the policies are just the same as other options like `RetryOptions` and `TelemetryPolicyOptions`. If we want to prevent customers from modifying policies, why don't we need to prevent them from modifying other options?

File: sdk/storage/azure-storage-blobs/src/blob_client.cpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-blobs/src/blob_client.cpp:56]: The order of storage policies is delicate sometimes very tricky. I appreciate your help in making things work with the new client options. But I think it'd be better to undo these changes and let storage developer take care of it.

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:118]: what does the `&&` after `auto` here mean?

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:49]: >If we want to prevent customers from modifying policies

What we want is to encapsulate the feature of `policies extensibility`.  The feature allow customers to add custom policies to the pipeline by calling `AddPolicy()` from any `clientOptions`. 

The problem with making the vector<policies> public is that the implementation details of the feature becomes the exposed and never to be changed. We might want to change the vector<> for something else (the implementation) and the way to do it is by encapsulating.  The customers just calls AddPolicy and set if they want it per call or per retry only.

Now, we can think about the policies as 3 types:
- Default: The ones mandated by Azure Core (like retry, telemetry, transport, etc). This policies ensure all SDK clients offers the same functionalities to customers. This policies are added by Azure Core and set it up through PolicyOptions as input. The policy options are in the ClientOptions base class.

- Service's policies:  These are specific to services and works together with the default ones. Same as the default policies, the customers never interact with the policy but with the policyOptions only. The options for this policies are in the clientOptions' derived class (like blobClientOptions).  Azure Core will never know about these options. The Service will use the options from the derived class to create the policies and will pass them to Azure Core to use them next to the default.

- Customer policies. Policies created by a customer and attached through the AddPolicy() API exposed by the clientOptions base class. The services will never know about these policies.

The new constructor for the Http pipeline takes the 3 types of policies to build the Pipeline for the service client.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:49]: okay, although I don't like this way. But it doesn't hurt me anyway.

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:118]: forward reference. `policy` will be a direct access to `perCallPolicies` instead, for instance, of creating a new reference `Policy&` to the lvalue.  
@CaseyCarter  did what I say make any sense?  🤣

File: sdk/storage/azure-storage-blobs/src/blob_client.cpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-blobs/src/blob_client.cpp:56]: Yeah, I can undo.
I wanted to provide some hints on how to call the new HttpPipeline constructor. But yes, please do not consider it as the final and desired fix to storage.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/src/blob_client.cpp:56]: Your `HttpPipepipeline` constructor won't work for storage scenario. Auth policy has to be the last one that can modify HTTP url or request headers.

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:118]: yeah, I get what you mean. But we all know `perCallPolicies` is const ref. So auto will be const lvalue reference. So
`for (auto& :)` or `for (const auto& :` will be enough. I just feel the extra `&` is kind of misleading.

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:43]: just for the record, I still think

```
struct ClientOptions {
    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perOperationPolicies;
    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perRetryPolicies;
    Azure::Core::Http::TransportPolicyOptions Transport;
    ...
};
```

would be better.

Other client options are gonna inherit from this class. A struct with member functions looks kind of weird

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:80]: what's the point of `std::mov`ing from a const?

File: CMakeLists.txt
Comment (vhvb1989) [reply CMakeLists.txt:96]: No. I have removed the fixes I did before for blobs as you requested.
Please fix and enable it back

Comment (Jinming-Hu) [reply CMakeLists.txt:96]: why it cannot be built? You only added something, you didn't remove anything or make any breaking changes.

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:43]: But this is not struct. This is a class which has an encapsulated way to AddPolicies()

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:80]: My bad, I will make the arg &&.  
The servicies will create and allocate the vector of policies and we just want to move the vector

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:43]: yeah, so `KeyClientOptions` is also going to be a `class` maybe? I dont know.

I also wrote some struct with member functions. Then someone (I cannot remember who) jumped out and said I cannot do that.

File: CMakeLists.txt
Comment (vhvb1989) [reply CMakeLists.txt]: Yes, you are right. 
I have enabled it back.
I would like to remove the HttpPipeline constructor which takes the list of policies. That would be the breaking change

Comment (Jinming-Hu) [reply CMakeLists.txt]: > I would like to remove the HttpPipeline constructor which takes the list of policies. That would be the breaking change

You mean this one https://github.com/Azure/azure-sdk-for-cpp/blob/27943f0819e76eb74eaf8c884b5a93dad4f608f5/sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp#L162?

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:43]: I think you would do the same for any storage client option which you would like to keep apart from its implementation details.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:43]: I don't like that and I don't think that'd be necessary and I don't see that'd benefit our customers. But I will make the change to be consistent with core sdk.

File: CMakeLists.txt
Comment (vhvb1989) [reply CMakeLists.txt]: Yes, that constructor is a way to workaround the mandated policies from Azure Core.
This was basically the arch board feedback and the reason of this entire PR.

The arch board is concerned about allowing services to choose all the policies for the pipeline, because they might decided not to add policies which Core wants to have in all SDK clients.

Comment (vhvb1989) [reply CMakeLists.txt]: Well, the arch board also mentioned that in the future, if it happen that we have 20 o 40 SDK services for C++, and we want to add a new client option to all of them, we want to do it only from Core and have all services updated.

File: sdk/storage/azure-storage-blobs/src/blob_client.cpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-blobs/src/blob_client.cpp]: I have updated the order to let service's policies to execute after custom-end-user policies.
That should be enough. (The logging policy at the end won't update the request. It only read from it).

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:118]: client policies should go after service policies, to enable customers to overwrite some default request headers

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:148]: client policies should go after service policies, to enable customers to overwrite some default request headers

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:148]: need another parameter in this function for auth policy

File: CMakeLists.txt
Comment (Jinming-Hu) [reply CMakeLists.txt]: okay, i'm sorry for making you undo the changes to storage policy construction code. It's just I knew it wouldn't work when I saw your changes, mainly because auth policy has to be the last policy like I mentioned in another thread, and retry policy to secondary problem which we already have a solution. I just want to make sure everything, every details of this `ClientOptions` is thoroughly considered before we make changes to those stuff depend on it.

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp]: Done

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (vhvb1989) [inline sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:109]: @Jinming-Hu  use this policy for auth

**Comment by Jinming-Hu** [comment]
Please hold this PR until https://github.com/Azure/azure-sdk-for-cpp/pull/1738 is merged or have an alternative for it

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:109]: This works for me

**Comment by vhvb1989** [comment]
remove extra overload and auth policy

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:108]: If you want encapsulation at all like you said in this comment https://github.com/Azure/azure-sdk-for-cpp/pull/1734/files#r582551593, why do you return const ref here? You expose the internal impl anyway

Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:108]: This is internal,  (if you ask me, I would move it to be private and grant access to this function only to the httpPipeline class to keep it as a detail imp, but Rick prefer not to use friend access).

So, yes, this method is making a promise to the public about returning the list of policies as a vector<Policy> which we will never break.
It does not, however, make it part of the contract that the pipeline internally is a vector<Policy>.

This is not too bad because the `AddPolicy` takes a Policy, and you can get the list of policies we add with `GetPolicies`...

**Comment by vhvb1989** [comment]
@Jinming-Hu  I requested your review again because I updated the name of the policy options and impacted storage. This rename changes are based on: https://apiview.dev/Assemblies/Review/3bd21beb2d9941a78683ac117c7bae33#Azure::Core::Http::RetryOptions

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:108]: > So, yes, this method is making a promise to the public about returning the list of policies as a vector which we will never break.

How? if as you said we change the internal impl from a vector to something else, we cannot return a const ref vector anymore for this function. This looks like a breaking change to me.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:108]: >  (if you ask me, I would move it to be private and grant access to this function only to the httpPipeline class to keep it as a detail imp, but Rick prefer not to use friend access).

If you want to use private and friend class, what's the point of `AddPolicy` and `GetPolicies` anyway? I can just access the private members

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:108]: I admit I don't like this encapsulation idea in the first place. But if you insist, i'm fine with it. It just seems to me that your explanation of encapsulation doesn't hold water

Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:108]: > How? if as you said we change the internal impl from a vector to something else, we cannot return a const ref vector anymore for this function. This looks like a breaking change to me.

The per call and per retry policies will be inside the clientOptions and will always be kept as a vector<Policy>.

The HttpPipeline will get them and use it to build a pipeline. The pipeline can be implemented as we want.  The clientOptions are inmutable.  The current implementation, for example, makes a clone of each policy from the clientOptions.  This ensure that the same clientOptions can be used to construct multiple clients, and each client will allocate its own pipeline.  Then we can dispose the clientOptions and the clients are still valid.
Then, for storage, a client can create a new client and re-use the same pipeline (shared_ptr).  The preferred way to create blobClients, for instance, is to spawn them from a containerClient instead of manually creating them from a clientOptions.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:108]: okay, although I still don't think it's a good design. We cannot get anything if we keep arguing.

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/http/policy.hpp:408]: `ValueOptions`? doesn't sound like a descriptive name

**Review by Jinming-Hu — APPROVED** [review]
Hi @vhvb1989 , thanks for your work! All looks good to me, except definition of `ClientOptions`, about which we've had a lot of discussions, and apparently neither of us can convince each other. Maybe we can ask for @JeffreyRichter 's opinion.

Hi @JeffreyRichter , I think the definition of `ClientOptions` in [this](https://github.com/Azure/azure-sdk-for-cpp/blob/fd097ed4336cb0a2d6ff771e4d446f6a004e3583/sdk/core/azure-core/inc/azure/core/internal/client_options.hpp) file can be simplified. We can make it a struct and have per-operation policies and per-retry policies public. Then we don't need `AddPolicy` or `GetPerCallPolicies` or `GerPerRetryPolicies` functions anymore, which accomplish nothing but a getter and `std::vector<>::push_back`. Then we can also get rid of `enum HttpPolicyOrder` [here](https://github.com/Azure/azure-sdk-for-cpp/blob/fd097ed4336cb0a2d6ff771e4d446f6a004e3583/sdk/core/azure-core/inc/azure/core/http/policy.hpp#L37).

@vhvb1989 You can merge this PR first. I cannot adapt the new `ClientOptions` in storage yet because retry counter in context is not ready yet. I'll do it after retry counter and context are ready.

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:109]: This last option parameter is unnessecary.  Storage can pass in the perRetryPolicies and put their authBeforeSendPolicy as the last in that list.  That is what other languages do, why isn't that acceptable here?

**Review by RickWinter — CHANGES_REQUESTED** [review]
Can we remove the authBeforeSend and have Storage pass in auth as part of a perretry array?  Other languages do not have this special parameter.  I don't see a reason that C++ needs it either

**Comment by JeffreyRichter** [comment]
If the goal is that client libraries expose ClientOptions and derive off of it (and I DO think that this is the goal), then:

*                 ClientOptions should be in the Azure::Http namespace (not in Internal)

*                 The per-operation & per-retry policy vectors should be public because customer can define/construct their own policy types and insert them into the pipeline.

*                 We do not need AddPolicy/GetPerCallPolicies/GerPerRetryPolicies methods.

*                 Making it a struct is good as all members should be public


  *   Jeffrey Richter (see Architecting Distributed Cloud Apps<http://aka.ms/RichterCloudApps> & Designing HTTP APIs<https://www.youtube.com/watch?v=9Ng00IlBCtw&list=PL9XzOCngAkqs4m0XdULJu_78nM3Ok3Q65&index=1>)

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (JeffreyRichter) [inline sdk/core/azure-core/inc/azure/core/http/policy.hpp:37]: I don't think we need this if we remove AddPolicy method

File: sdk/core/azure-core/inc/azure/core/http/policy.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/policy.hpp:408]: I have created an issue to remove this policy from Core #1767 .

The reason that policy was created is to help the protocol implementation details from an sdk service.  An sdk client should maintain a policy like such but Coro should now encourage that

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:75]: The parameters are for the sdk service client to create a pipeline.

The client is exposing `telemetryOptions` to the customer, which only contains `ApplicationId`.

The TelemetryPolicy takes `TelemetryPolicyOptions` plus the `serviceName` and the `ServiceVersion`.  These last two parameters are provided by the service client.

This parameters is the way Azure Core is making sure that all services would add a telemetryPolicy

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:77]: The customer's custom policies are inside the clientOptions base client.  They come here as part of the clientOptions struct

These other other parameters are the service's custom policies

File: sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp:109]: @Jinming-Hu , as per Azure Core team conversation (and also Ted and Pavel from .Net), we will not support the feature to make sure the addition of a policy to be right before the transport policy.

It looks like in all the history of storage .Net, they have never faced the issue of an end-user adding a per-retry policy which updates the request combined with the `sharedKeyAuthPolicy`.  They mention that end-users typically uses per-call policies only.

File: sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp
Comment (vhvb1989) [reply sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:28]: The current approach is to make `ClientSecretCredentialOptions` to have a `TransportPolicy`, and all other policies (like retry) are hardcoded.
This policy will be creating a pipeline and sending a request, it makes sense to allow end-users to define how that request would handle.

Comment (RickWinter) [reply sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:28]: Thinking out loud. The ClientSecretCredentailOptions could have a constructor that takes in the base client options.

**Review by RickWinter — APPROVED** [review]

**Comment by vhvb1989** [comment]
> ClientOptions should be in the Azure::Http namespace (not in Internal) 

It's internal because we expect only service's clients to derive from it to create client Options.  If we make it public, it means that as an end-user, I can use for something.

> The per-operation & per-retry policy vectors should be public because customer can define/construct their own policy types and insert them into the pipeline.

Fine. I think there is some value to keep them hidden. For example, in a multi-thread scenario where the client options can be updated for one thread while another is creating a client with it. We could just let AddPolicy() and GetPolicies() take a mutex.  But with no abstraction, we will just not support some scenarios like that.  Let me make that change as part of a new PR.

**Comment by vhvb1989** [comment]
@JeffreyRichter 

There are 2 things I will follow up after this PR is merge that you mentioned:
- Should ClientSecretCredentialOptions derive from clientOptions?
- Should per-call and per-retry end-user's policies array be public?

**Pull request merged**
